File: app/utils/defoldTweens/Actions.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayReduce = ____lualib.__TS__ArrayReduce
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local ____exports = {}
local _createActions
local ____Easings = require("utils.defoldTweens.Easings")
local Easings = ____Easings.default
function _createActions(obj)
    local getOpt, _startTween, defaultOptions, thisTween
    function getOpt(opt, key)
        local ____temp_1
        if opt ~= nil and opt[key] ~= nil then
            ____temp_1 = opt[key]
        else
            ____temp_1 = defaultOptions[key]
        end
        return ____temp_1
    end
    function _startTween(easingFunction, from, to, time, callback)
        local timeElapsed = 0
        timer.delay(
            0.016,
            true,
            function(_, handler, timeElapsedFromTrigger)
                timeElapsed = timeElapsed + timeElapsedFromTrigger
                if timeElapsed >= time then
                    callback(
                        easingFunction(time, from, to - from, time),
                        true
                    )
                    timer.cancel(handler)
                    return
                end
                callback(easingFunction(timeElapsed, from, to - from, time))
            end
        )
    end
    local sequence = {}
    defaultOptions = {
        easing = "linear",
        onUpdate = function(____, currentValue, toValue)
        end,
        onComplete = function(____, currentValue)
        end
    }
    local function to(prop, value, duration, opt)
        sequence[#sequence + 1] = {
            key = "TO",
            prop = prop,
            duration = duration,
            delay = 0,
            val = value,
            opt = opt
        }
        return thisTween
    end
    local function start()
        if #sequence > 0 then
            __TS__ArrayReduce(
                sequence,
                function(____, acc, tween, idx)
                    local ____tween_0 = tween
                    local prop = ____tween_0.prop
                    local duration = ____tween_0.duration
                    local val = ____tween_0.val
                    local opt = ____tween_0.opt
                    local easing = getOpt(opt, "easing")
                    local onUpdate = getOpt(opt, "onUpdate")
                    local onComplete = getOpt(opt, "onComplete")
                    _startTween(
                        Easings[easing],
                        obj[prop],
                        val,
                        duration,
                        function(val, isFinished)
                            obj[prop] = val
                            onUpdate(nil, obj[prop], val)
                            if isFinished then
                                onComplete(nil, obj[prop])
                            end
                        end
                    )
                    acc = acc + duration
                    return acc
                end,
                0
            )
            __TS__ArraySplice(sequence, 0, #sequence)
        end
    end
    thisTween = {to = to, start = start}
    return thisTween
end
____exports.action = _createActions
return ____exports

File: src/utils/defoldTweens/Actions.ts
/* eslint-disable @typescript-eslint/no-non-null-assertion */
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unnecessary-type-assertion */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
import { ExtractPropsType } from "../types";
import Easings, { EasingFunction } from "./Easings";

export type Actions = ReturnType<typeof _createActions>;


export const action = _createActions;

function _createActions<T extends object>(obj: T) {
    type Props = keyof ExtractPropsType<T, number>;

    type Options = { easing?: keyof typeof Easings, onUpdate?: (currentValue: number, toValue: number) => void, onComplete?: (currentValue: number) => void };
    type ToTween = { key: 'TO', prop: Props, duration: number, delay: number, val: number, opt?: Options };
    type TweenData = ToTween;

    const sequence: TweenData[] = [];
    const defaultOptions: Required<Options> = {
        easing: 'linear',
        onUpdate: (currentValue: number, toValue: number): void => { },
        onComplete: (currentValue: number): void => { },
    };

    function to<P extends Props>(prop: P, value: number, duration: number, opt?: Options) {
        sequence.push({ key: 'TO', prop, duration, delay: 0, val: value, opt });
        return thisTween;
    }

    function start() {
        if (sequence.length > 0) {
            sequence.reduce((acc, tween, idx) => {
                const { prop, duration, val, opt } = tween;

                const easing = getOpt(opt, 'easing');
                const onUpdate = getOpt(opt, 'onUpdate')!;
                const onComplete = getOpt(opt, 'onComplete')!;

                _startTween(Easings[easing!] as EasingFunction, obj[prop] as number, val, duration, (val, isFinished) => {
                    (obj as any)[prop] = val;

                    onUpdate(obj[prop] as number, val);
                    if (isFinished)
                        onComplete(obj[prop] as number);
                });

                acc += duration;

                return acc;
            }, 0);
            sequence.splice(0, sequence.length);
        }
    }

    function getOpt<keys extends keyof Options>(opt: Options | undefined, key: keys): Options[keys] {
        return opt != undefined && opt[key] != undefined ? opt[key]! : defaultOptions[key];
    }

    function _startTween(easingFunction: EasingFunction, from: number, to: number, time: number, callback: (value: number, finished?: boolean) => void): void {
        let timeElapsed = 0;

        timer.delay(0.016, true, (_: any, handler: hash, timeElapsedFromTrigger: number) => {
            timeElapsed += timeElapsedFromTrigger;
            if (timeElapsed >= time) {
                callback(easingFunction(time, from, to - from, time), true);
                timer.cancel(handler);
                return;
            }
            callback(easingFunction(timeElapsed, from, to - from, time));
        });
    }

    const thisTween = {
        to,
        start
    };

    return thisTween;
}